source/agora/consensus/EnrollmentManager.d

```diff
@@ -474,6 +474,16 @@ public class EnrollmentManager
 
     public void clearExpiredValidators (Height block_height) @safe nothrow
     {
+        // clear up the enrollment of a node if the validator cycle of the node
+        // ends at the `block_height`
+        const enrolled = this.validator_set.getEnrolledHeight(this.enroll_key);
```

Comment:
It might be useful to move this into a private function called `isEnrolled()`. Then inside of `getPreimage` you could add an assert: `assert(!this.isEnrolled());` to catch programming bugs.